### PR TITLE
Refactor realtime-communication and replace y-protocols and y-codemirror.next

### PR DESCRIPTION
**Pull request by mrdrogdrog**
### Component/Part
Realtime communication

### Description

This PR:
- refactors the whole realtime communication to work with json instead of binary
- introduces event based message processing
- removes y-protocols
- removes y-codemirror.next and replaces it with an own plugin

### Steps

- [x] Added implementation
- [x] Added / updated tests
- [x] Added / updated documentation
- [x] I read the [contribution documentation](https://github.com/hedgedoc/hedgedoc/blob/develop/CONTRIBUTING.md) and
  made sure that:
  - My commits are signed-off to accept the DCO.
  - This PR targets the correct branch: `master` for 1.x & docs, `develop` for 2.x

### Related Issue(s)
Replaces #3270 because codecov didn't trigger correctly.

### CODE REVIEW

File: backend/src/realtime/realtime-note/realtime-user-status.spec.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+  let realtimeNote: RealtimeNote;
+
+  beforeEach(() => {
+    realtimeNote = new RealtimeNote(
+      Mock.of<Note>({ id: 9876 }),
+      'mockedContent',
+    );
+    client1 = new MockConnectionBuilder(realtimeNote)
+      .withRealtimeUserState()
+      .withUsername('mock1')
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply backend/src/realtime/realtime-note/realtime-user-status.spec.ts]: Why? They're not used anywhere else.

**Comment by mrdrogdrog** [comment]
Things that could be improved:
- Save/Cache the Y-Doc before deleting to preserve the structures and make re-syncs easier
- Cooldown before the realtime doc gets deleted

File: frontend/src/components/editor-page/editor-pane/codemirror-extensions/remote-cursors/cursor-colors.module.scss [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+.cursor-5 {
+  --color: #ffff00;
+}
+
+.cursor-6 {
+  --color: #00fff2;
+}
+
+.cursor-7 {
```
[the remark this answers is not in the document]
Comment (mrdrogdrog) [reply frontend/src/components/editor-page/editor-pane/codemirror-extensions/remote-cursors/cursor-colors.module.scss:35]: It's difficult to find more than 8 distinct colors. I would leave it like this for now and add more later.

File: frontend/src/components/editor-page/websocket-connection-modal/realtime-connection-modal.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+import { useApplicationState } from '../../../hooks/common/use-application-state'
+import { WaitSpinner } from '../../common/wait-spinner/wait-spinner'
+import React from 'react'
+import { Col, Container, Modal, Row } from 'react-bootstrap'
+import { Trans, useTranslation } from 'react-i18next'
+
+export const RealtimeConnectionModal: React.FC = () => {
```
Comment (ErikMichelson) [inline frontend/src/components/editor-page/websocket-connection-modal/realtime-connection-modal.tsx:12]:
```suggestion
/**
 * Modal with a spinner that is shown while reconnecting to the realtime backend
 */
export const RealtimeConnectionModal: React.FC = () => {
```

File: frontend/src/components/editor-page/editor-pane/hooks/yjs/use-y-doc-sync-client-adapter.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import type { Listener } from 'eventemitter2'
+import { useEffect, useMemo } from 'react'
+import type { Doc } from 'yjs'
+
+const logger = new Logger('useYDocSyncClient')
+
+/**
+ * Creates a {@link YDocSyncClientAdapter} and mirrors its sync state to the global application state.
+ *
+ * @param messageTransporter The {@link MessageTransporter message transporter} that sends and receive messages for the synchronisation
```
Comment (ErikMichelson) [inline frontend/src/components/editor-page/editor-pane/hooks/yjs/use-y-doc-sync-client-adapter.ts:19]:
```suggestion
 * @param messageTransporter The {@link MessageTransporter message transporter} that sends and receives messages for the synchronisation
```

**Review by ErikMichelson — CHANGES_REQUESTED** [review]
Only two docs things, nothing else left from my side

**Pull request merged**